design/004-IoT-Adoption-Watcher.md

```diff
@@ -0,0 +1,248 @@
+state was not the latest off-chain state, the watcher must immediately register
+the latest off-chain state.  Given the connectivity and power constraints of
+IoT devices, it might not be possible to meet these requirements if watching
+service is running on the IoT device.
+
+The current design cannot be extended to implement a watcher that can run
+as a remote service. Hence, a new design is proposed where, the watcher can be run:
+1. As a part of go-perun client itself (ideal for capable hardware).
+2. As an independent component on the same or different computer (ideal for
+   constrained hardware).
+
+## Details
+
+<!-- Provide a detailed description of the proposal. -->
+
+### Functionalities of the watcher
+
+1. User should be able to initialize the watcher without associating it to any
+   particular channel. After initialization, it must have access to a wallet
+   containing keys for an on-chain account with sufficient funds (to pay
+   transaction fees when registering off-chain states).
+
+2. Main component should be able to register/deregister channel IDs to
+   start/stop watching for on-chain events.
+
+3. After a channel is registered,
+
+   1. Main component should be able to periodically send the newer off-chain
+      states.
+
+   2. Main component should be notified by the watcher, when a state is
+      registered or progressed on the blockchain.
+
+4. If the state registered on the blockchain and is not the latest off-chain
+   state known to the watcher, then it must register the latest state on the
+   blockchain.
+
+### Interfaces of the watcher
+
+In order to realize the desired functionalities, watcher should provide the
+following interfaces:
+
+
+```go
+type Watcher struct{}
+
+// For initalizing the watcher.
+func NewWatcher(rs RegisterSubscriber) *Watcher {...}
+
+// For interacting with the blockchain.
+type RegisterSubscriber interface {
+    Subscriber
+    Registerer
+}
+
+// For registering off-chain state for a given ledger channel,
+// along with the off-chain states of its sub-channels and virtual channels.
+type Registerer interface {
+    Register(State, SubStates)
+}
+
+// For receiving registered and progressed events for a given channel ID.
+type Subscriber interface {
+    Subscribe(ChannelID)
+}
+
+// For registering/de-registering a channel ID with/from the watcher.
+func (w *Watcher) Register(ChannelID, OffChainStatesSub, OnChainEventsPub) {...}
+func (w *Watcher) Deregister(ChannelID) {...}
+
+// For receiving off-chain states from the main component.
+type OffChainStatesSub interface {
+	Next() OffChainState
+}
+
+// For sending on-chain events to the main component.
+type OnChainEventsPub inteface {
+    // OnChainEvent can be a registered event or a progressed event.
+    Publish(OnChainEvent)
+}
+
+// Interfaces for the main component to interact with the watcher.
+
+// For receiving on-chain events from the watcher.
+type OnChainEventsSub interface {
```

Comment:
Where and how is this used?

Reply:
Have updated comments for this and all other interfaces to describe their usage.

Reply:
I have re-ordered the sections by moving interface definition after the interaction diagram. Because, the interface definition is derived from the interaction diagram.

Do you think it is easy to relate the two now ?